IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -554,33 +528,42 @@ - (MSIDCredentialCacheItem *)tokenWithKey:(MSIDCacheKey *)key
                                               context:(id<MSIDRequestContext>)context
                                                 error:(NSError **)error
 {
-    NSArray<MSIDCredentialCacheItem *> *tokenList = @[];
+    NSArray *itemList = @[];
     
-
     /*
-     Sync in memory cache with persistent cache at the time of look up.
+     For refresh tokens, always merge with persistence to get the most recent refresh token as it is shared across apps from same publisher.
+     For AT/ID tokens, two apps sharing the same client id can write to the same entry to the keychain. In this case, it is possible that the first app is trying to read a credential which is not currently in its own memory but is written in persistence by the second app sharing the same client id. To find this credential, it is important to merge in memory cache with persistence.
      */
-    MSIDMacCredentialStorageItem *savedItem = [self storageItemWithKey:key serializer:serializer context:context error:error];
-    MSIDMacCredentialStorageItem *currentItem = key.isShared ? self.sharedStorageItem : self.appStorageItem;
     
-    if (!key.isShared)
+    MSIDMacCredentialStorageItem *storageItem = key.isShared ? self.sharedStorageItem : self.appStorageItem;
+    MSIDMacCredentialStorageItem *savedStorageItem = [self storageItemWithKey:key serializer:serializer context:context error:error];
+    
+    if ([key isKindOfClass:[MSIDDefaultCredentialCacheKey class]])
     {
-        /*
-         For refresh tokens, always merge with persistence to get the most recent refresh token as it is shared across apps from same publisher.
-         For AT/ID tokens, two apps sharing the same client id can write to the same entry to the keychain. In this case, it is possible that the first app is trying to read a credential which is not currently in its own memory but is written in persistence by the second app sharing the same client id. To find this credential, it is important to merge in memory cache with persistence.
-         */
-        tokenList = [currentItem storedCredentialsForKey:(MSIDDefaultCredentialCacheKey *)key];
-        if ([tokenList count])
+        MSIDDefaultCredentialCacheKey *cacheKey = (MSIDDefaultCredentialCacheKey *)key;
+        NSString *bucket = [self getBucketForCredentialType:cacheKey.credentialType];
```

Comment:
we seem to have two places where bucket is decided/used - here to get bucket based on credential type and in the in-memory layer to create cache item based on the bucket. Would it make sense to have a single place for it, e.g. in memory cache?

Reply:
Agree. Moved logic to get item type from cacheKey instead of credential to MSIDMacCredentialStorage class. Used cacheKey as it is the only thing common in api's saveToken, tokensWithKey and removeTokensWithKey.